File: src/submission_handler.rs
use anyhow::Context;
use worker::{Request, Response, RouteContext, Url};

use crate::alerting::send_alert;
use crate::embed_generator::default_open_graph::generate_default_opengraph_embed;
use crate::embed_generator::generic_embed_open_graph::generate_generic_opengraph_embed;
use crate::embed_generator::message_open_graph::generate_message_opengraph_embed;
use crate::embed_generator::telegram_embed_open_graph::generate_telegram_opengraph_embed;
use crate::furaffinity::submission_info::SubmissionInfoResponse;
use crate::{furaffinity::client::FurAffinity, requester::Requester};

pub async fn handle_submission(
    req: Request,
    context: RouteContext<FurAffinity>,
) -> Result<Response, worker::Error> {
    let result = handle_submission_inner(req, &context).await;

    if let Err(e) = &result {
        send_alert(&context.env, &format!("{:?}", e)).await;
    }

    match result {
        Ok(resp) => Ok(resp),
        Err(_) => Ok(Response::from(generate_default_opengraph_embed())),
    }
}

async fn handle_submission_inner(
    req: Request,
    context: &RouteContext<FurAffinity>,
) -> anyhow::Result<Response> {
    let submission_id = context.param("id").unwrap().parse();
    if submission_id.is_err() {
        return Ok(Response::from(generate_message_opengraph_embed(
            "User Error",
            "Please use a valid submission URL to generate an embed.",
        )));
    }
    let submission_id = submission_id.unwrap();

    let requester = Requester::from_user_agent(
        req.headers()
            .get("user-agent")
            .unwrap()
            .unwrap_or_default()
            .as_str(),
    );

    if let Requester::Human = requester {
        return Ok(Response::redirect(
            Url::parse(&context.data.get_submission_url(submission_id)).unwrap(),
        )
        .unwrap());
    }

    let info_response = context
        .data
        .fetch_submission_info(submission_id)
        .await
        .with_context(|| format!("Failed to fetch submission info for {submission_id}"))?;

    let embed = match info_response {
        SubmissionInfoResponse::ImageSubmission(info) => match requester {
            Requester::Telegram => generate_telegram_opengraph_embed(&info)?,
            Requester::OtherBot => generate_generic_opengraph_embed(&info)?,
            Requester::Human => generate_generic_opengraph_embed(&info)?,
        },
        SubmissionInfoResponse::FlashSubmission => generate_message_opengraph_embed(
            "Unsupported Submission",
            "Flash content cannot be shown as a preview",
        ),
        SubmissionInfoResponse::NotFound => generate_message_opengraph_embed(
            "Not Found",
            &format!("The submission {submission_id} was not found on FurAffinity"),
        ),
        SubmissionInfoResponse::ServerError => generate_message_opengraph_embed(
            "FA Down",
            "FurAffinity responded with a server error, which means it's probably down at the moment, or encountered an error"
        ),
    };

    Ok(Response::from(embed))
}

File: src/submissionHandler.ts
import { getCached, setCached } from './cache.js';
import type { Config } from './config.js';
import { fetchSubmissionInfo } from './furaffinity/client.js';
import type { ServerErrorDetail } from './furaffinity/submissionInfo.js';
import { generateImageEmbed, generateImageTelegramEmbed } from './embedGenerator/imageEmbed.js';
import { generateMessageEmbed } from './embedGenerator/messageEmbed.js';
import { generateMusicEmbed, generateMusicTelegramEmbed } from './embedGenerator/musicEmbed.js';
import { generateStoryEmbed } from './embedGenerator/storyEmbed.js';
import { noticeError } from './metrics.js';
import { classifyRequester } from './requester.js';

export type RequestMeta = {
  requester: 'human' | 'telegram' | 'otherBot';
  cached: boolean | null;
  submissionResult: string | null;
  serverError?: ServerErrorDetail;
  error?: unknown;
};

export type HandlerResult =
  | { type: 'redirect'; url: string; meta: RequestMeta }
  | { type: 'embed'; html: string; meta: RequestMeta };

export async function handleSubmission(
  id: number,
  userAgent: string,
  config: Config
): Promise<HandlerResult> {
  const requester = classifyRequester(userAgent);

  if (requester === 'human') {
    return { type: 'redirect', url: `https://www.furaffinity.net/view/${id}`, meta: { requester, cached: null, submissionResult: null } };
  }

  try {
    const cached = await getCached(config.cacheDir, id);
    const result = cached ?? await fetchSubmissionInfo(id, { a: config.sessionA, b: config.sessionB });
    if (!cached) await setCached(config.cacheDir, id, result);

    const meta: RequestMeta = { requester, cached: cached !== null, submissionResult: result.type };

    switch (result.type) {
      case 'image': {
        const oEmbedUrl = config.publicUrl ? `${config.publicUrl}/oembed?id=${id}` : undefined;
        const html =
          requester === 'telegram'
            ? generateImageTelegramEmbed(result.info)
            : generateImageEmbed(result.info, oEmbedUrl);
        return { type: 'embed', html, meta };
      }
      case 'story': {
        const oEmbedUrl = config.publicUrl ? `${config.publicUrl}/oembed?id=${id}` : undefined;
        return { type: 'embed', html: generateStoryEmbed(result.info, oEmbedUrl), meta };
      }
      case 'music': {
        const oEmbedUrl = config.publicUrl ? `${config.publicUrl}/oembed?id=${id}` : undefined;
        const html =
          requester === 'telegram'
            ? generateMusicTelegramEmbed(result.info)
            : generateMusicEmbed(result.info, oEmbedUrl);
        return { type: 'embed', html, meta };
      }
      case 'flash':
        return { type: 'embed', html: generateMessageEmbed('Unsupported Submission', 'Flash content cannot be shown as a preview'), meta };
      case 'notFound':
        return { type: 'embed', html: generateMessageEmbed('Not Found', `The submission ${id} was not found on FurAffinity`), meta };
      case 'serverError':
        return {
          type: 'embed',
          html: generateMessageEmbed('FA Down', 'FurAffinity responded with a server error, which means it\'s probably down at the moment, or encountered an error'),
          meta: { ...meta, serverError: result.detail },
        };
      case 'unauthenticated':
        return { type: 'embed', html: generateMessageEmbed('Session Expired', "FurAffinity has invalidated xfuraffinity's session, please try again later."), meta };
      case 'blocked':
        return { type: 'embed', html: generateMessageEmbed('Blocked by FurAffinity', 'FurAffinity is blocking automated access'), meta };
    }
  } catch (err) {
    noticeError(err);
    return {
      type: 'embed',
      html: generateMessageEmbed('xfuraffinity Error', 'An unexpected error occurred. Please report this at github.com/FirraWoof/xfuraffinity'),
      meta: { requester, cached: null, submissionResult: null, error: err },
    };
  }
}
